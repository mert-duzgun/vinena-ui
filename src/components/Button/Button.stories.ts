import type { Meta, StoryObj } from '@storybook/react';

import { Button } from './Button';

// More on how to set up stories at: https://storybook.js.org/docs/7.0/react/writing-stories/introduction
const meta = {
  title: 'Button',
  component: Button,
  tags: ['autodocs'],
} satisfies Meta<typeof Button>;

export default meta;
type Story = StoryObj<typeof meta>;

// More on writing stories with args: https://storybook.js.org/docs/7.0/react/writing-stories/args
export const Primary: Story = {
  args: {
    priority: 'primary',
    children: 'Button',
  },
};

export const Secondary: Story = {
  args: {
    priority: 'secondary',
    children: 'Button',
  },
};

export const Large: Story = {
  args: {
    size: 'large',
    children: 'Button',
  },
};

export const Small: Story = {
  args: {
    size: 'small',
    children: 'Button',
  },
};

export const Disabled: Story = {
  args: {
    priority: 'primary',
    children: 'Button',
    disabled: true,
  },
};

export const PrimaryWithIcon: Story = {
  args: {
    priority: 'primary',

    children: 'Button',
  },
};

export const SecondaryWithIcon: Story = {
  args: {
    priority: 'secondary',

    children: 'Button',
  },
};
