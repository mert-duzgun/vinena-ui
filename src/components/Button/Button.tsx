import clsx from 'clsx';
import Icon from 'components/Icon/Icon';
import { AnchorHTMLAttributes, ButtonHTMLAttributes } from 'react';

interface ButtonProps {
  variant?: 'primary' | 'secondary';
  as?: 'button' | 'a';
  size?: 'small' | 'default' | 'large';
  className?: string;
  icon?: string;
}

export const Button = ({
  children,
  className,
  variant = 'primary',
  as = 'button',
  size = 'default',
  icon,
  ...rest
}: ButtonProps &
  ButtonHTMLAttributes<HTMLButtonElement> &
  AnchorHTMLAttributes<HTMLAnchorElement>) => {
  let baseStyle, textSize, iconStyle;
  switch (size) {
    case 'small':
      baseStyle = 'px-4 py-2';
      textSize = 'text-sm';
      iconStyle = 'text-[16px]';
      break;
    case 'large':
      baseStyle = 'px-4 py-3.5';
      textSize = 'text-lg';
      iconStyle = 'text-[24px]';
      break;
    default:
      baseStyle = 'px-4 py-3';
      textSize = 'text-base';
      iconStyle = 'text-[20px]';
  }
  const baseStyleClasses = `${baseStyle} inline-flex items-center border transition rounded focus:outline-none focus:ring-[3px] focus:ring-primary-300`;
  let colorStyle;
  switch (variant) {
    case 'primary':
      colorStyle =
        'text-white bg-primary-500 border-primary-500 hover:bg-primary-700 hover:border-primary-700 disabled:bg-grayscale-300 disabled:border-grayscale-300';
      break;
    case 'secondary':
      colorStyle =
        'text-primary-500 bg-transparent border-primary-500 hover:bg-primary-50 disabled:text-grayscale-300 disabled:border-grayscale-300 disabled:bg-transparent';
      break;
  }

  const classes = clsx(baseStyleClasses, textSize, colorStyle, className);

  if (as === 'a') {
    return (
      <a className={clsx(classes, 'cursor-pointer')} {...rest}>
        {icon ? <Icon name={icon} className={iconStyle} /> : null}
        <span className={clsx({ 'ml-[10px]': icon })}>{children}</span>
      </a>
    );
  } else {
    return (
      <button className={classes} type="button" {...rest}>
        {icon ? <Icon name={icon} className={iconStyle} /> : null}
        <span className={clsx({ 'ml-[10px]': icon })}>{children}</span>
      </button>
    );
  }
};
