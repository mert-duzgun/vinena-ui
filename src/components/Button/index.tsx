export function Button() {
  return <button>Button</button>;
}
