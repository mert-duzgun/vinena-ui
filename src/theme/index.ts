export * from './GlobalStyles';
export * from './theme';
export * from './utils';
