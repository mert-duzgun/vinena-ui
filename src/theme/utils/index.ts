export * from './px/px';
export * from './rem/rem';
