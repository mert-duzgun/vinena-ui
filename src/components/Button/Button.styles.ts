import { px, rem, VinenaTheme } from 'theme';

const sizes = {
  small: {
    height: rem(32),
    paddingTop: rem(4),
    paddingRight: rem(16),
    paddingBottom: rem(4),
    paddingLeft: rem(16),
  },
  default: {
    height: rem(40),
    paddingTop: rem(4),
    paddingRight: rem(16),
    paddingBottom: rem(4),
    paddingLeft: rem(16),
  },
  large: {
    height: rem(48),
    paddingTop: rem(4),
    paddingRight: rem(24),
    paddingBottom: rem(4),
    paddingLeft: rem(24),
  },
};

const getFonts = (theme: VinenaTheme) => {
  return {
    small: theme.fonts.p3,
    default: theme.fonts.p2,
    large: theme.fonts.p1,
  };
};

const getPriorityStyles = (theme: VinenaTheme) => {
  return {
    primary: {
      backgroundColor: theme.colors.primary[500],
      borderColor: theme.colors.primary[500],
      color: theme.colors.white,

      '&:hover': {
        backgroundColor: theme.colors.primary[300],
        borderColor: theme.colors.primary[300],
      },

      '&:active': {
        backgroundColor: theme.colors.primary[700],
        borderColor: theme.colors.primary[700],
      },

      '&:disabled': {
        backgroundColor: theme.colors.grayscale[300],
        borderColor: theme.colors.grayscale[300],
      },

      '&[data-negative="true"]': {
        backgroundColor: theme.colors.danger[500],
        borderColor: theme.colors.danger[500],

        '&:hover': {
          backgroundColor: theme.colors.danger[400],
          borderColor: theme.colors.danger[400],
        },

        '&:active': {
          backgroundColor: theme.colors.danger[700],
          borderColor: theme.colors.danger[700],
        },

        '&:disabled': {
          backgroundColor: theme.colors.grayscale[400],
          borderColor: theme.colors.grayscale[400],
        },
      },
    },
    secondary: {
      backgroundColor: theme.colors.white,
      borderColor: theme.colors.primary[500],
      color: theme.colors.primary[500],

      '&:hover': {
        backgroundColor: theme.colors.primary[50],
        borderColor: theme.colors.primary[500],
      },

      '&:active': {
        backgroundColor: theme.colors.primary[100],
        borderColor: theme.colors.primary[700],
      },

      '&:disabled': {
        backgroundColor: theme.colors.white,
        borderColor: theme.colors.grayscale[300],
        color: theme.colors.grayscale[300],
      },

      '&[data-negative="true"]': {
        backgroundColor: theme.colors.white,
        borderColor: theme.colors.danger[500],
        color: theme.colors.danger[500],

        '&:hover': {
          backgroundColor: theme.colors.danger[50],
          borderColor: theme.colors.danger[500],
        },

        '&:active': {
          backgroundColor: theme.colors.danger[100],
          borderColor: theme.colors.danger[700],
        },

        '&:disabled': {
          backgroundColor: theme.colors.white,
          borderColor: theme.colors.grayscale[400],
          color: theme.colors.grayscale[400],
        },
      },
    },
    tertiary: {
      backgroundColor: theme.colors.white,
      borderColor: theme.colors.transparent,
      color: theme.colors.primary[500],
      textDecoration: 'underline',

      '&:hover': {
        backgroundColor: theme.colors.primary[50],
        borderColor: theme.colors.primary[50],
      },

      '&:active': {
        backgroundColor: theme.colors.primary[100],
        borderColor: theme.colors.primary[100],
        color: theme.colors.primary[700],
      },

      '&:disabled': {
        backgroundColor: theme.colors.white,
        borderColor: theme.colors.white,
        color: theme.colors.grayscale[300],
      },

      // No negative variant for tertiary because a destructive action should be at least secondary
    },
  };
};

export interface ButtonStylesParams {
  priority: 'primary' | 'secondary' | 'tertiary';
  size: 'small' | 'default' | 'large';
  negative: boolean;
  disabled: boolean;
}

export default function createStyles(theme: VinenaTheme, props: ButtonStylesParams) {
  return {
    root: {
      minHeight: rem(32),
      borderRadius: px(9999),
      cursor: props.disabled ? 'not-allowed' : 'pointer',
      transition: 'all 0.2s',
      borderWidth: px(1),
      borderStyle: 'solid',
      display: 'inline-flex',
      alignItems: 'center',
      fontFamily: 'Inter, sans-serif',
      textDecoration: 'none',
      ...sizes[props.size],
      ...getFonts(theme)[props.size],
      ...getPriorityStyles(theme)[props.priority],

      '&:focus-visible': {
        outlineStyle: 'solid',
        outlineWidth: px(2),
        outlineOffset: px(2),
        outlineColor:
          props.negative && props.priority !== 'tertiary'
            ? theme.colors.danger[700]
            : theme.colors.primary[300],
      },
    },
  };
}
