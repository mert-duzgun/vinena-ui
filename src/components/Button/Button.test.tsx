import { ThemeProvider } from '@emotion/react';
import { fireEvent, render, screen } from '@testing-library/react';
import { DEFAULT_THEME, rem } from 'theme';
import { vitest } from 'vitest';

import { Button } from './Button';

describe('Button', () => {
  it('should match snapshot', async () => {
    const { container } = render(
      <ThemeProvider theme={DEFAULT_THEME}>
        <Button>Click me!</Button>
      </ThemeProvider>
    );
    await expect(container.firstChild).toMatchSnapshot();
  });

  it('renders with default props', () => {
    render(
      <ThemeProvider theme={DEFAULT_THEME}>
        <Button>Click me!</Button>
      </ThemeProvider>
    );

    expect(screen.getByRole('button', { name: 'Click me!' })).toBeInTheDocument();
  });

  it('renders with a custom component prop', () => {
    render(
      <ThemeProvider theme={DEFAULT_THEME}>
        <Button as="a" href="https://duzgun.dev">
          Click me!
        </Button>
      </ThemeProvider>
    );

    expect(screen.getByRole('link', { name: 'Click me!' })).toHaveAttribute(
      'href',
      'https://duzgun.dev'
    );
  });

  it('calls onClick handler when clicked', () => {
    const handleClick = vitest.fn();
    render(
      <ThemeProvider theme={DEFAULT_THEME}>
        <Button onClick={handleClick}>Click me!</Button>
      </ThemeProvider>
    );

    fireEvent.click(screen.getByRole('button', { name: 'Click me!' }));
    expect(handleClick).toHaveBeenCalled();
  });

  it('renders with disabled state', () => {
    render(
      <ThemeProvider theme={DEFAULT_THEME}>
        <Button disabled>Click me!</Button>
      </ThemeProvider>
    );
    const button = screen.getByRole('button', { name: 'Click me!' });
    expect(button).toBeDisabled();
    expect(button).toHaveStyle('cursor: not-allowed');
  });

  it('renders negative state', () => {
    render(
      <ThemeProvider theme={DEFAULT_THEME}>
        <Button negative>Click me!</Button>
      </ThemeProvider>
    );
    const button = screen.getByRole('button', { name: 'Click me!' });

    expect(button).toHaveStyle(`background-color: ${DEFAULT_THEME.colors.danger[500]}`);
  });

  it('renders small default button', () => {
    render(
      <ThemeProvider theme={DEFAULT_THEME}>
        <Button size="small">Click me!</Button>
      </ThemeProvider>
    );
    const button = screen.getByRole('button', { name: 'Click me!' });

    expect(button).toHaveStyle(`height: ${rem(32)}`);
  });

  it('renders large default button', () => {
    render(
      <ThemeProvider theme={DEFAULT_THEME}>
        <Button size="large">Click me!</Button>
      </ThemeProvider>
    );
    const button = screen.getByRole('button', { name: 'Click me!' });

    expect(button).toHaveStyle(`height: ${rem(48)}`);
  });

  it('renders small default button', () => {
    render(
      <ThemeProvider theme={DEFAULT_THEME}>
        <Button size="small">Click me!</Button>
      </ThemeProvider>
    );
    const button = screen.getByRole('button', { name: 'Click me!' });

    expect(button).toHaveStyle(`height: ${rem(32)}`);
  });
});
