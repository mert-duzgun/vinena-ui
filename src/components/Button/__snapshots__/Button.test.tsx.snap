// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button > should match snapshot 1`] = `
<button
  class="css-1iwb2b9-Button"
  data-negative="false"
>
  <span>
    Click me!
  </span>
</button>
`;
